Rebase on Presto 307

BenoitHanotte opened the pull request:
Presto 307 contains the new `MetastoreLocator` API implemented to allow discovering the metastore using consul.
+ Change to throttle split discovery so that we can deploy & test in preprod

zhenxiao committed:
Add max data size per query

Signed-off-by: Raphael CLAUDE <r.claude@criteo.com>

piyushnarang committed:
Fix issue: 10099 - create table failed when using viewfs

Signed-off-by: Raphael CLAUDE <r.claude@criteo.com>

piyushnarang committed:
Criteo light travis yaml

Purpose is to avoid timeout & waisting cpu cycles on unused/unmodified
features
* Comment out some database connectors modules
* Comment out SqlServer docker tests
* Turn off travis slack updates

piyushnarang commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/DiscoveryHiveCluster.java:
```diff
@@ -0,0 +1,103 @@
+        ServiceHealth chosen = response.get(0);
+        String host = chosen.getNode().getNode();
+        int port = chosen.getService().getPort();
+        log.info("Connecting to metastore %s:%d", host, port);
```
debug for this too?

szehonCriteo replied:
So in my experience this kind of logs is pretty useful information when we debug hive connection issues, I'm afraid if we turn this to debug then it will just go to trash.

piyushnarang replied:
It seems like the metastore is created pretty often so based on my limited understanding this would be a fair bit of calls. If it doesn't look too spammy based on your testing we can leave it in.

piyushnarang commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/DiscoveryHiveClusterConfig.java:
```diff
@@ -0,0 +1,57 @@
+
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
```
this license isn't needed right? just the one at the top

szehonCriteo replied:
Will fix

piyushnarang commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/DiscoveryHiveClusterConfig.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Config("hive.metastore.uri")
+    @ConfigDescription("Consul URI of the form consul://consul-host:consul-port/service-name")
+    public DiscoveryHiveClusterConfig setConsulUri(String uri)
```
do we also need the `hive.metastore.username` ? we currently don't use it. But it's there in the static config. Do you think it's needed / useful?

szehonCriteo replied:
Nope, this should be picked up from kerberos ticket I think.

piyushnarang commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/DiscoveryHiveCluster.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    @Override
+    public HiveMetastoreClient createMetastoreClient()
```
so one potential perf issue with the current implementation is that we make a call to consul everytime `createMetastoreClient` is called. In the old implementation as the list was static this call just shuffled and returned a host. This might prove to be a slightly hot code path as it looks like createMetastoreClient is called a lot - https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java (pretty much on all Hive calls). I'm wondering if we should lookup the consul hosts, cache them for a few mins and refresh after some time. Maybe cache for 5-10 mins?

szehonCriteo replied:
But I think we need the most up date information to avoid query failures if the hive metastore moves, and also consul has a localhost agent which we will use that is supposed to return pretty quickly and do some caching.

piyushnarang replied:
I'll check that out. I'm pretty concerned about the performance as making a service call to keep looking up hosts seems like overkill. Especially when the frequency with which these updates would happen is likely once every few hours or every day or so. An option might also be to cache for fewer minutes too (or refresh on any connection failure).

szehonCriteo replied:
ok, i implemented a 3 min cache.

szehonCriteo commented on pom.xml:
```diff
@@ -700,6 +700,12 @@
+            <dependency>
+                <groupId>com.orbitz.consul</groupId>
+                <artifactId>consul-client</artifactId>
+                <version>1.1.1</version>
```
There are some in a property and some aren't and I didn't see the pattern when to do it, but will fix.

szehonCriteo commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/DiscoveryHiveCluster.java:
```diff
@@ -0,0 +1,103 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
```
Yes, will fix

szehonCriteo commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/DiscoveryHiveCluster.java:
```diff
@@ -0,0 +1,103 @@
+        String service = consulURI.getPath().substring(1);  //strip leading slash
+        int consulPort = consulURI.getPort();
+
+        // check that agentHost has scheme or not
```
Will fix comment

szehonCriteo commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/DiscoveryHiveCluster.java:
```diff
@@ -0,0 +1,103 @@
+        // check that agentHost has scheme or not
+        checkArgument(!isNullOrEmpty(consulHost), "Unspecified consul host, please use consul://consul-host:consul-port/service-name");
+        checkArgument(consulPort != -1, "Unspecified consul port, please use consul://consul-host:consul-port/service-name");
+        checkArgument(!isNullOrEmpty(service), "Unspecified consul service, please use consul://consul-host:consul-port/service-name");
```
Will fix to use the service argument, thanks for the catch.

piyushnarang commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/DiscoveryHiveCluster.java:
```diff
@@ -58,56 +59,69 @@ public DiscoveryHiveCluster(DiscoveryHiveClusterConfig config, HiveMetastoreClie
         this.clientFactory = clientFactory;
         this.unresolvedUris = config.getMetastoreUris();
+        //Trigger first resolution to fail fast.
+        resolvedUriSupplier.get();
```
this fails at Presto startup? or the first hive command that we issue? If it fails at startup that might be a problem in case of intermittent consul problems right?

piyushnarang commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/thrift/TestDiscoveryHiveCluster.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    @Test
+    public void testMetastoreFailedWithoutFallbackWithConsul()
```
maybe one test to add is the positive test for consul. Where the consul client actually gets returned.

piyushnarang commented:
Seems like one of the test failures is related to this change:
```
Caused by: com.facebook.presto.spi.PrestoException: Failed connecting to Hive metastore: [thrift://hadoop-master:9083]
	at com.facebook.presto.hive.metastore.thrift.DiscoveryHiveCluster.createMetastoreClient(DiscoveryHiveCluster.java:91)
	at com.facebook.presto.hive.metastore.thrift.ThriftHiveMetastore.lambda$getTable$5(ThriftHiveMetastore.java:206)
	at com.facebook.presto.hive.metastore.thrift.HiveMetastoreApiStats.lambda$wrap$0(HiveMetastoreApiStats.java:42)
	at com.facebook.presto.hive.RetryDriver.run(RetryDriver.java:137)
	at com.facebook.presto.hive.metastore.thrift.ThriftHiveMetastore.getTable(ThriftHiveMetastore.java:205)
	at com.facebook.presto.hive.metastore.thrift.BridgingHiveMetastore.getTable(BridgingHiveMetastore.java:77)
	at com.facebook.presto.hive.metastore.CachingHiveMetastore.loadTable(CachingHiveMetastore.java:255)```

Commit:
Throttle Hive splits discovery from the HiveSplitSource

Add `hive.max-splits-per-sec` configuration option to allow throttling
the number of splits released by the HiveSplitSource to provide strong
guarantee on the rate of read operations on the namenode.

Use Guava's RateLimiter to simplify code

Set hive.max-splits-per-second to Integer.MAX_VALUE by default

qinghui-xu committed:
Query history extension

* PR(origin) of query history extension

This is the original PR for creating query history extension.

The architecture is:
(Read part)
________                                    __________
                     ______________
|        |--- request for a query's info -->|          |---- QueryId
----->|              |
| web ui |                                  | REST api |
                   | QueryManager |
|________|<-- Serialized QueryInfo (json) --|__________|<---- QueryInfo
---|______________|

^              |

|              |

QueryInfo       QueryId

|              |

|              |
____                   ______________
                                     |______________v
|    |----- Json ----->|              |-- Deserialize json to QueryInfo
-->|              |
| DB |                 | HistoryStore |
                                    | QueryTracker |
|____|<-- SQL Query ---|______________|<------- request for QueryId
-------|______________|

(write part)
___________________                       ______________
                                 ____
|   QueryTracker    |                     |              |
                               |    |
| (event when query |-- Save QueryInfo -->| HistoryStore |--Serialize
QueryInfo (json) ->| DB |
| finished)         |                     |              |
                               |    |
|___________________|                     |______________|
                               |____|

* Plug QueryHistoryStore directly to REST API

So that we can directly send json string back to client web ui, instead
of having to reparse the string into some complex java object which is
error prone.
Also, this should save some cpu cycles by removing the json
deserialization/serialization overhead.

By doing this, we also remove some customized json serialization logic.

* Add index to the query history table

Remove the test with h2 because the create table with index syntax is
unsupported.

* Keep the QueryTracker signature

To be easier for future upstream rebase.

* QueryHistoryStore architecture

* [Checkstyle] Fix an error

* Put the query-history-extension.properties under etc dir

Commit:
Discover meastore URIs using consul

Add the DisoveryMetastoreLocator to resolve the metastore uri from
consul.
